### PR TITLE
improve addr/inv trickle logic

### PR DESCRIPTION
**Pull request by pstratem**
The addr/msg_tx logic currently assumes that ProcessMEssage/SendMessage are called on a regular timer.

This was (sort of) true in the past but is not longer true after #5971

Critical analysis of this PR would be helpful as this goes beyond just fixing the timing issue.

### CODE REVIEW

**Comment by jgarzik** [comment]
Any possibility to use the timer stuff being added in #5964 ?

**Comment by pstratem** [comment]
@jgarzik maybe later, porting wont be difficult

File: src/main.cpp [hunk 1]
```diff
@@ -4529,31 +4516,24 @@ bool SendMessages(CNode* pto, bool fSendTrickle)
...
-
         //
         // Message: addr
         //
-        if (fSendTrickle)
-        {
+
+        if (pto->nNextLocalAddrSend < GetTime()) {
+            AdvertizeLocal(pto);
+            pto->nNextLocalAddrSend = GetTime() + GetRand(24 * 60 * 60);
```
Comment (Diapolo) [inline src/main.cpp]: Could that magic number be defined somehwere as constant? IMHO it's used quite often in this pull.

Comment (pstratem) [inline src/main.cpp]: At a later date in a different PR

File: src/main.cpp [hunk 2]
```diff
@@ -4619,54 +4600,72 @@ bool SendMessages(CNode* pto, bool fSendTrickle)
...
+        vInv.reserve(1000);
+        
+        if (pto->nNextInvSend < nNowMicros)
+            vInvWait.reserve(pto->vInventoryToSend.size());
 
-                    if (fTrickleWait)
+        {
+            LOCK(pto->cs_inventory);
+            
+            // TODO it would be sufficient to walk vInventoryToSend randomly, shuffling is relatively inefficient
```
Comment (sipa) [inline src/main.cpp]: In a random walk, it is hard to efficiently avoid processing the same element twice. If shuffling is slow, create an array of indexes, and shuffle that. Also, where does std::random_shuffle() get its randomness?

Suggested code:
```
std::vector<int> vRandInv;
vRandInv.resize(pto->vInventoryToSend.size());
for (int i = 0; i < pto->vRandInv.size(); i++) vRandInv[i] = i;
for (int i = 0; i < pto->vRandInv.size(); i++) std::swap(vRandInv[i], vRandInv[i + GetRand(vRandInv.size() - i]);

for (int i = 0; i < pto->vRandInv.size(); i++) {
    const CInv& inv = pto->vInventoryToSend[vRandInv[i]];

    ...
```

File: src/main.cpp [hunk 3]
```diff
@@ -4619,54 +4600,72 @@ bool SendMessages(CNode* pto, bool fSendTrickle)
...
         if (!vInv.empty())
             pto->PushMessage("inv", vInv);
 
+        if (pto->nNextInvSend < nNowMicros)
+            pto->nNextInvSend = nNowMicros + GetRand(10 * 1000000);
+
         // Detect whether we're stalling
-        int64_t nNow = GetTimeMicros();
-        if (!pto->fDisconnect && state.nStallingSince && state.nStallingSince < nNow - 1000000 * BLOCK_STALLING_TIMEOUT) {
+        nNowMicros = GetTimeMicros();
```
Comment (sipa) [inline src/main.cpp]: You already have an initialized nNowMicros variable. No need to request the current time again?

File: src/main.cpp [hunk 4]
```diff
@@ -4619,54 +4600,72 @@ bool SendMessages(CNode* pto, bool fSendTrickle)
...
+            std::random_shuffle(pto->vInventoryToSend.begin(), pto->vInventoryToSend.end());
+
+            BOOST_FOREACH(const CInv& inv, pto->vInventoryToSend) {
+                switch(inv.type) {
+                    case MSG_BLOCK:
+                    case MSG_FILTERED_BLOCK:
                     {
-                        vInvWait.push_back(inv);
-                        continue;
+                        // Aggressively push MSG_BLOCK/MSG_FILTERED_BLOCK
```
Comment (sipa) [inline src/main.cpp]: Maybe it is more efficient to keep a separate vInventoryToSend for MSG_BLOCK and MSG_FILTERED_BLOCK, and only shuffle/process the normal one when nNextInvSend is exceeded.

**Comment by sipa** [comment]
Concept ACK. I love getting rid of the vSendTrickle logic...

File: src/main.cpp (hunk 3 above)
Comment (pstratem) [inline src/main.cpp]: There's a certain amount of error from caching the value.

The value is updated twice to reduce this error.

Comment (sipa) [inline src/main.cpp]: If the difference in time between them is more than a few milliseconds, we have different problems :)

Comment (pstratem) [inline src/main.cpp]: I want to avoid changing the logic of anything im not explicitly touching.

Comment (sipa) [inline src/main.cpp]: Fair enough.

**Comment by gavinandresen** [comment]
I'm with @jgarzik: this could be much cleaner built on top of #5964. All of the pnode->nNext* variables could go away; just schedule a task that takes a NodeId and Does The Right Thing at some (random) time in the future.

(the tasks would lock cs_vNodes and do the NodeId to CNode* lookup... probably don't need to optimize that lookup, looping through even a 1,000-peer vector isn't a significant amount of CPU time).

**Comment by sipa** [comment]
Unsure. I don't think running the scheduled operations makes sense if there
is nothing else to do for a node, and there almost always is, at which time
we already hold the necessary locks.

Small benefit, small cost.

**Comment by gavinandresen** [comment]
@sipa : benefit would be clearer logic. I find the "set a variable here, test it over there and then do something" hard to review/follow.

**Comment by pstratem** [comment]
@gavinandresen As I said to @jgarzik it's easy enough to back port this logic into CSchedule.

I would rather not tie the two PR's together.

File: src/main.cpp [hunk 5]
```diff
@@ -4553,31 +4540,25 @@ bool SendMessages(CNode* pto, bool fSendTrickle)
...
         //
         // Message: addr
         //
-        if (fSendTrickle)
-        {
+
+        int64_t nNowMicros = GetTimeMicros();
+        if (pto->nNextLocalAddrSend < nNowMicros) {
+            AdvertizeLocal(pto);
+            pto->nNextLocalAddrSend = nNowMicros + GetRand(24 * 60 * 60 * 1000000);
```
Comment (sipa) [inline src/main.cpp]: Integer overflow on this line. You probably want to cast to int64_t first.

**Comment by sipa** [comment]
@gavinandresen Yes, that's the small benefit. The small cost is extra locking/work. Not opposed to it, but it's not as clear an advantage as replacing the several pieces of code we had that were already using their own thread. This one doesn't need it.

**Comment by sipa** [comment]
Can you fix the integer overflow, and perhaps rebase on top of #6066?

**Comment by sipa** [comment]
Rebase please?

**Comment by sipa** [comment]
Needs rebase, and fixing the integer overflow.

**Comment by sipa** [comment]
Did a rebase and fix: https://github.com/sipa/bitcoin/commits/trickle_fix

**Comment by sipa** [comment]
@pstratem ping

**Comment by pstratem** [comment]
@sipa pong

**Comment by pstratem** [comment]
setAddrKnown was replaced with a CRollingBloomFilter d81cff32e50fe5f686f985d0af2e74219f328ed0

Given this it seems reasonable to slowly unset bits of the bloomfilter rather than clearing the filter entirely every 24 hours.

Thoughts?

**Comment by pstratem** [comment]
Scratch that. There's no need to clear the filter explicitly with the rolling bloom filter.

**Comment by dcousens** [comment]
@pstratem rebase/close?

**Comment by sipa** [comment]
rebase please...

**Comment by sipa** [comment]
See #7125.

**Comment by pstratem** [comment]
we have better ideas on how to achieve the properties of the trickle logic (such as mempool sync at random interval)

File: src/main.cpp [hunk 6]
```diff
@@ -4860,54 +4841,91 @@ bool SendMessages(CNode* pto, bool fSendTrickle)
...
 
-                    if (fTrickleWait)
+            for (size_t i = 0; i < vRandInv.size(); i++)
+                std::swap(vRandInv[i], vRandInv[i + GetRand(vRandInv.size() - i)]);
+
+            for (size_t i = 0; i < vRandInv.size(); i++) {
+                const CInv& inv = pto->vInventoryToSend[vRandInv[i]];
+
+                switch(inv.type) {
+                    case MSG_TX:
```
Comment (rebroad) [inline src/main.cpp]: surely no need to check this since only TXs are pushed to this vector

**Pull request closed without merging**